presto-main/src/main/java/com/facebook/presto/catalogserver/RandomCatalogServerAddressSelector.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.catalogserver;
+
+import com.facebook.drift.client.address.AddressSelector;
+import com.facebook.presto.metadata.InternalNodeManager;
+import com.facebook.presto.spi.HostAddress;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.net.HostAndPort;
+
+import javax.inject.Inject;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.function.Function;
+
+import static com.facebook.drift.client.address.SimpleAddressSelector.SimpleAddress;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+public class RandomCatalogServerAddressSelector
```

Comment:
Not necessarily since the part that differers is here

>         List<HostAndPort> catalogServers = internalNodeManager.getCatalogServers().stream()

so there is no differing type between the two classes. Since they are both of type Set<InternalNode>. Their boolean "isResourceManager", etc in the InternalNode class will be what differs